### Phase 1 — Data Understanding

The objective of this phase is to understand the structure, scope, and quality
of the raw dataset before making any changes to the data.

In [1]:
import numpy as np
import pandas as pd

#### 1. Load and Inspect the Dataset

- In this section, we load the raw e-commerce dataset and perform an initial inspection to understand its size, structure, data types, and sample records.
- No data is modified during this phase.

In [13]:
data = pd.read_csv("../data/raw/online_retail_II.csv")
print(data.shape)
data.head()

(1067371, 8)


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [14]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


#### 2. Data Quality Check

In [16]:
data.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

In [19]:
data.duplicated().sum()

34335

In [48]:
temp = data.copy()
temp['InvoiceDate'] = pd.to_datetime(temp['InvoiceDate'])
temp['InvoiceDate'].min()

Timestamp('2009-12-01 07:45:00')

In [49]:
temp['InvoiceDate'].max()

Timestamp('2011-12-09 12:50:00')

In [50]:
temp['InvoiceDate'].describe()

count                          1067371
mean     2011-01-02 21:13:55.394028544
min                2009-12-01 07:45:00
25%                2010-07-09 09:46:00
50%                2010-12-07 15:28:00
75%                2011-07-22 10:23:00
max                2011-12-09 12:50:00
Name: InvoiceDate, dtype: object

In [28]:
print(data.nunique())

Invoice        53628
StockCode       5305
Description     5698
Quantity        1057
InvoiceDate    47635
Price           2807
Customer ID     5942
Country           43
dtype: int64


In [29]:
data['Quantity'].describe()

count    1.067371e+06
mean     9.938898e+00
std      1.727058e+02
min     -8.099500e+04
25%      1.000000e+00
50%      3.000000e+00
75%      1.000000e+01
max      8.099500e+04
Name: Quantity, dtype: float64

In [30]:
(data['Quantity'] < 0).sum()

22950

In [31]:
(data['Quantity'] == 0).sum()

0

In [32]:
data['Price'].describe()

count    1.067371e+06
mean     4.649388e+00
std      1.235531e+02
min     -5.359436e+04
25%      1.250000e+00
50%      2.100000e+00
75%      4.150000e+00
max      3.897000e+04
Name: Price, dtype: float64

In [34]:
(data['Price']<=0).sum()

6207

In [37]:
data['Invoice'].head(10)

0    489434
1    489434
2    489434
3    489434
4    489434
5    489434
6    489434
7    489434
8    489435
9    489435
Name: Invoice, dtype: object

In [38]:
data[data['Invoice'].astype(str).str.startswith('C')].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia


In [39]:
data['Invoice'].astype(str).str.startswith('C').sum()

19494

In [42]:
data['StockCode'].value_counts().head(10)

StockCode
85123A    5829
22423     4424
85099B    4216
21212     3318
20725     3259
84879     2960
47566     2768
21232     2747
22197     2549
22383     2540
Name: count, dtype: int64

In [46]:
data['StockCode'].unique()[:50]

array(['85048', '79323P', '79323W', '22041', '21232', '22064', '21871',
       '21523', '22350', '22349', '22195', '22353', '48173C', '21755',
       '21754', '84879', '22119', '22142', '22296', '22295', '22109',
       '22107', '22194', '35004B', '82582', '21181', '21756', '21333',
       '84596F', '84596L', '22111', '22143', '22145', '22130', '21364',
       '21360', '21351', '21352', '35400', '20695', '37370', '10002',
       '84507B', '20703', '21987', '21989', '84970S', '20971', '22271',
       '22272'], dtype=object)

In [47]:
data['StockCode'].nunique()

5305

### Phase 1 — Initial Findings

- Dataset contains 1067371 rows and 8 columns.
- 2 columns contain missing values.
- 34335 duplicate rows were found.
- Dataset covers the period from 2009-12-01 to 2011-12-09.
- There are 47635 unique invoices.
- There are 5305 unique products.
- There are 5942 unique customers.
- There are 43 countries.
- Quantity contains 22950 negative values and 0 'zero' values.
- Price contains 6207 zero/negative values.
- 19494 transaction records have invoice numbers beginning with `C` and may represent cancellations; this will be investigated during data cleaning.
- Other suspicious observations: `InvoiceDate` is stored as an object, `Customer ID` is stored as float64, and several invoice/product codes
  require further investigation before cleaning..